folding-schemes/src/folding/protogalaxy/circuits.rs

```diff
@@ -88,34 +107,363 @@ impl FoldingGadget {
 
         let e_star = F_alpha * &L_X_evals[0] + Z_X.evaluate(&gamma)? * K_X.evaluate(&gamma)?;
 
-        let mut u_star = &instance.u * &L_X_evals[0];
         let mut x_star = instance.x.mul_scalar(&L_X_evals[0])?;
         for i in 0..k {
-            u_star += &vec_instances[i].u * &L_X_evals[i + 1];
             x_star = x_star.add(&vec_instances[i].x.mul_scalar(&L_X_evals[i + 1])?)?;
         }
 
         // return the folded instance
-        Ok(CommittedInstanceVar {
-            betas: betas_star,
-            // phi will be computed in CycleFold
-            phi: NonNativeAffineVar::new_constant(ConstraintSystemRef::None, C::zero())?,
-            e: e_star,
-            u: u_star,
-            x: x_star,
-        })
+        Ok((
+            CommittedInstanceVar {
+                betas: betas_star,
+                // phi will be computed in CycleFold
+                phi: NonNativeAffineVar::new_constant(ConstraintSystemRef::None, C::zero())?,
+                e: e_star,
+                x: x_star,
+            },
+            L_X_evals,
+        ))
+    }
+}
+
+pub struct AugmentationGadget {}
+
+impl AugmentationGadget {
+    #[allow(clippy::type_complexity)]
+    pub fn prepare_and_fold_primary<C: CurveGroup, S: CryptographicSponge>(
+        transcript: &mut impl TranscriptVar<CF1<C>, S>,
+        U: CommittedInstanceVar<C>,
+        u_phis: Vec<NonNativeAffineVar<C>>,
+        u_xs: Vec<Vec<FpVar<CF1<C>>>>,
```

Comment:
For this and https://github.com/privacy-scaling-explorations/sonobe/pull/123#discussion_r1736253307, I intentionally declared the parameters as `Vec` for the same reason as https://github.com/privacy-scaling-explorations/sonobe/pull/123#discussion_r1748943011, where both methods internally need owned vector elements.

So if they are slices, we need to clone their elements inside the methods, while in the current implementation, no cloning is required if the caller doesn't need to own the arguments anymore.

I agree with you that the caller will need to allocate vectors, which is not ideal. Maybe a better solution is to have fixed-size arrays as parameters, although this would make the code more verbose IMO.

Reply:
Ohh I see.
The issue is [how the internals are defined](
https://github.com/privacy-scaling-explorations/sonobe/blob/52de2d185c67acf33977bcaf567fb8321191c4b9/folding-schemes/src/folding/circuits/nonnative/uint.rs#L177).

Yea, it's sad we need to actually do this...
I assume this can be improved. But agree to leave it like this now.